mautic/mautic
# Avoid using requestStack->getCurrentRequest() in constructor

| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | features
| Bug fix?                               | yes
| New feature?                           | no
| Deprecations?                          | no
| BC breaks?                             | no
| Automated tests included?              | N/A
| Related user documentation PR URL      | N/A
| Related developer documentation PR URL | N/A
| Issue(s) addressed                     | Fixes #... 




#### Description:

https://github.com/mautic/mautic/pull/9544 introduced a regression in the API Library, specifically in the CategoryModel: `Call to a member function get() on null`.

This is because the `$requestStack->getCurrentRequest()` is already called in the constructor. Due to how Dependency Injection works, it might be that `getCurrentRequest()` returns `null` if the function is called too early. That PR surfaced this "bug".

Looking at [Symfony's official docs](https://symfony.com/doc/current/service_container/request.html), they also don't call `getCurrentRequest()` in the constructor already, but later on in the consuming function.

I came across several other places in Mautic that have this behavior; also fixed that in this PR.

**Before applying this PR** (https://github.com/mautic/api-library/runs/1895155796?check_suite_focus=true):

```
Running phpunit in 2 processes with /home/runner/work/api-library/api-library/vendor/phpunit/phpunit/phpunit

Configuration read from /home/runner/work/api-library/api-library/phpunit.xml.dist

........................S.......FFFFFF.......S.................  63 / 304 ( 20%)
.................................S............................. 126 / 304 ( 41%)
............................................................... 189 / 304 ( 62%)
.....................................................S......... 252 / 304 ( 82%)
..............SS....S...............................            304 / 304 (100%)

Time: 08:32.874, Memory: 18.00 MB

There were 6 failures:

1) Mautic\Tests\Api\CategoriesTest::testGetList
The response has unexpected status code (500).

Response: {"errors":[{"message":"Call to a member function get() on null","code":500,"type":null}],"trace":[{"namespace":"","short_class":"","class":"","type":"","function":"","file":"\/var\/www\/html\/mautic\/app\/bundles\/CategoryBundle\/Model\/CategoryModel.php","line":55,"args":[]},{"namespace":"Mautic\\CategoryBundle\\Model","short_class":"CategoryModel","class":"Mautic\\CategoryBundle\\Model\\CategoryModel","type":"-\u003E","function":"getPermissionBase","file":"\/var\/www\/html\/mautic\/app\/bundles\/ApiBundle\/Controller\/CommonApiController.php","line":546,"args":[]},{"namespace":"Mautic\\ApiBundle\\Controller","short_class":"CommonApiController","class":"Mautic\\ApiBundle\\Controller\\CommonApiController","type":"-\u003E","function":"initialize","file":"\/var\/www\/html\/mautic\/app\/bundles\/CategoryBundle\/Controller\/Api\/CategoryApiController.php","line":31,"args":[]},{"namespace":"Mautic\\CategoryBundle\\Controller\\Api","short_class":"CategoryApiController","class":"Mautic\\Categ...

Failed asserting that false is true.

/home/runner/work/api-library/api-library/tests/Api/MauticApiTestCase.php:89
/home/runner/work/api-library/api-library/tests/Api/MauticApiTestCase.php:146
/home/runner/work/api-library/api-library/tests/Api/MauticApiTestCase.php:203
/home/runner/work/api-library/api-library/tests/Api/CategoriesTest.php:26

.... etc. etc.
```

**After applying this PR (api-library tests):**

```
$ vendor/bin/paratest tests/Api/CategoriesTest.php
Running phpunit in 8 processes with /var/www/html/vendor/phpunit/phpunit/phpunit

Configuration read from /var/www/html/phpunit.xml.dist

.......                                                             7 / 7 (100%)

Time: 00:36.165, Memory: 4.00 MB

OK (7 tests, 120 assertions)
```



#### Steps to test this PR:
1. Load up the API library locally
2. Load up this PR locally
3. Make sure the API library tests config (`tests/local.config.php`) points at your Mautic installation that has this PR applied; using Basic Auth on the API is easiest for testing
4. In the API library folder, run `vendor/bin/paratest tests/Api/CategoriesTest.php`. All tests should pass.
